File: backend/app/api/routes/listings.py
"""
app/api/routes/listings.py — مسارات الإعلانات
دورة الحياة: مسودة ← صور+AI ← وثائق ← سعر ← مراجعة ← نشر
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_, func, or_
from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel
import uuid

from app.core.database import get_db
from app.core.config import settings
from app.models.models import (
    Listing, ListingImage, ListingCategory,
    ListingType, ListingStatus, ConditionGrade, User
)
from app.api.routes.auth import get_current_user

router = APIRouter()


# ── Schemas ───────────────────────────────────────────────────────────────────

class CreateListingIn(BaseModel):
    category:     ListingCategory
    listing_type: ListingType
    title:        str
    description:  Optional[str] = None
    city:         Optional[str] = None
    district:     Optional[str] = None


class SetPriceIn(BaseModel):
    price: float


class UpdateDetailsIn(BaseModel):
    details: dict


# ── Helpers ───────────────────────────────────────────────────────────────────

def get_price_limit(category: str, listing_type: str, condition: str) -> float:
    try:
        return settings.PRICE_CAPS[category][listing_type][condition]
    except KeyError:
        return 99_999_999


def _serialize(l: Listing) -> dict:
    return {
        "id":               str(l.id),
        "category":         l.category.value,
        "listing_type":     l.listing_type.value,
        "title":            l.title,
        "description":      l.description,
        "status":           l.status.value,
        "city":             l.city,
        "district":         l.district,
        "price":            l.price,
        "price_max_limit":  l.price_max_limit,
        "currency":         l.currency,
        "condition_grade":  l.condition_grade.value if l.condition_grade else None,
        "condition_score":  l.condition_score,
        "condition_report": l.condition_report,
        "details":          l.details,
        "images":           [{"url": i.url, "type": i.image_type, "order": i.order}
                             for i in (l.images or [])],
        "seller_name":      l.seller.full_name if l.seller else None,
        "seller_id":        str(l.seller_id),
        "view_count":       l.view_count,
        "published_at":     l.published_at.isoformat() if l.published_at else None,
        "created_at":       l.created_at.isoformat(),
    }


async def _get_or_404(listing_id: str, db: AsyncSession) -> Listing:
    result  = await db.execute(select(Listing).where(Listing.id == uuid.UUID(listing_id)))
    listing = result.scalar_one_or_none()
    if not listing:
        raise HTTPException(404, "الإعلان غير موجود")
    return listing


# ── Step 1: Create draft ──────────────────────────────────────────────────────

@router.post("/", status_code=201)
async def create_listing(
    body: CreateListingIn,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    listing = Listing(
        id=uuid.uuid4(),
        seller_id=current_user.id,
        category=body.category,
        listing_type=body.listing_type,
        title=body.title,
        description=body.description,
        city=body.city,
        district=body.district,
        status=ListingStatus.draft,
    )
    db.add(listing)
    await db.commit()
    return {"listing_id": str(listing.id), "status": "draft"}


# ── Step 2: Receive condition result from upload route ────────────────────────

@router.post("/{listing_id}/set-condition")
async def set_condition(
    listing_id: str,
    grade: ConditionGrade,
    score: float,
    report: dict,
    db: AsyncSession = Depends(get_db),
):
    """يُستدعى داخلياً من upload route بعد تحليل AI."""
    listing = await _get_or_404(listing_id, db)
    listing.condition_grade  = grade
    listing.condition_score  = score
    listing.condition_report = report
    listing.price_max_limit  = get_price_limit(
        listing.category.value, listing.listing_type.value, grade.value
    )
    listing.status = ListingStatus.condition_assessed
    await db.commit()
    return {
        "condition_grade": grade.value,
        "condition_score": score,
        "price_max_limit": listing.price_max_limit,
        "currency":        listing.currency,
    }


# ── Step 3: Update extra details ──────────────────────────────────────────────

@router.patch("/{listing_id}/details")
async def update_details(
    listing_id: str,
    body: UpdateDetailsIn,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    listing = await _get_or_404(listing_id, db)
    if str(listing.seller_id) != str(current_user.id):
        raise HTTPException(403, "غير مصرح")
    listing.details = body.details
    await db.commit()
    return {"status": "updated"}


# ── Step 4: Set price ─────────────────────────────────────────────────────────

@router.post("/{listing_id}/set-price")
async def set_price(
    listing_id: str,
    body: SetPriceIn,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    listing = await _get_or_404(listing_id, db)
    if str(listing.seller_id) != str(current_user.id):
        raise HTTPException(403, "غير مصرح")

    if listing.status != ListingStatus.docs_verified:
        raise HTTPException(400, "يجب إتمام التحقق من الوثائق أولاً")

    if listing.price_max_limit and body.price > listing.price_max_limit:
        raise HTTPException(
            422,
            f"السعر يتجاوز الحد الأقصى ({listing.price_max_limit:,.0f} {listing.currency})"
        )

    listing.price  = body.price
    listing.status = ListingStatus.price_set
    await db.commit()
    return {"price": body.price, "status": "price_set"}


# ── Step 5: Submit for admin review ──────────────────────────────────────────

@router.post("/{listing_id}/submit")
async def submit_for_review(
    listing_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    listing = await _get_or_404(listing_id, db)
    if str(listing.seller_id) != str(current_user.id):
        raise HTTPException(403, "غير مصرح")
    if listing.status != ListingStatus.price_set:
        raise HTTPException(400, "يجب اكتمال جميع الخطوات")

    listing.status = ListingStatus.pending_review
    await db.commit()
    return {"status": "pending_review", "message": "تم إرسال الإعلان للمراجعة ✅"}


# ── Public: Browse ────────────────────────────────────────────────────────────

@router.get("/")
async def get_listings(
    category:     Optional[ListingCategory] = None,
    listing_type: Optional[ListingType]     = None,
    condition:    Optional[ConditionGrade]  = None,
    city:         Optional[str]  = Query(None),
    q:            Optional[str]  = Query(None),
    min_price:    Optional[float] = None,
    max_price:    Optional[float] = None,
    page:         int = Query(1, ge=1),
    page_size:    int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    filters = [Listing.status == ListingStatus.published]
    if category:     filters.append(Listing.category == category)
    if listing_type: filters.append(Listing.listing_type == listing_type)
    if condition:    filters.append(Listing.condition_grade == condition)
    if city:         filters.append(Listing.city.ilike(f"%{city}%"))
    if min_price:    filters.append(Listing.price >= min_price)
    if max_price:    filters.append(Listing.price <= max_price)
    if q:
        filters.append(or_(
            Listing.title.ilike(f"%{q}%"),
            Listing.description.ilike(f"%{q}%"),
        ))

    total = (await db.execute(
        select(func.count()).select_from(Listing).where(and_(*filters))
    )).scalar()

    rows = (await db.execute(
        select(Listing)
        .where(and_(*filters))
        .order_by(Listing.published_at.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
    )).scalars().all()

    return {
        "total":     total,
        "page":      page,
        "page_size": page_size,
        "pages":     (total + page_size - 1) // page_size,
        "items":     [_serialize(l) for l in rows],
    }


@router.get("/my")
async def get_my_listings(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """إعلاناتي الخاصة بجميع الحالات"""
    rows = (await db.execute(
        select(Listing)
        .where(Listing.seller_id == current_user.id)
        .order_by(Listing.created_at.desc())
    )).scalars().all()
    return [_serialize(l) for l in rows]


@router.get("/{listing_id}")
async def get_listing(listing_id: str, db: AsyncSession = Depends(get_db)):
    listing = await _get_or_404(listing_id, db)
    if listing.status != ListingStatus.published:
        raise HTTPException(404, "الإعلان غير متاح")
    listing.view_count = (listing.view_count or 0) + 1
    await db.commit()
    return _serialize(listing)

File: server/src/routes/listings.js
import { Router } from "express";
import { db, uid, json } from "../db.js";
import { requireAuth, isAdmin } from "../auth.js";
import { ai, computeTier } from "../ai.js";

export const router = Router();

function hydrate(row) {
  if (!row) return null;
  const images = db
    .prepare("SELECT url FROM listing_images WHERE listing_id = ? ORDER BY sort_order")
    .all(row.id)
    .map((r) => r.url);
  return {
    ...row,
    images,
    details: json(row.details, {}),
    condition_report: json(row.condition_report, null),
    docs_report: json(row.docs_report, null),
  };
}

function getOwned(id, user) {
  const row = db.prepare("SELECT * FROM listings WHERE id = ?").get(id);
  if (!row) return { error: [404, "الإعلان غير موجود — ابدأ إعلاناً جديداً"] };
  if (row.owner_id !== user.id && !isAdmin(user)) return { error: [403, "لا تملك صلاحية هذا الإعلان"] };
  return { row };
}

/** GET /listings/ — قائمة الإعلانات المنشورة مع الفلاتر */
router.get("/", (req, res) => {
  const { city, category, kind, listing_type, price_tier, q, condition_grade } = req.query;
  const min = req.query.min_price, max = req.query.max_price;
  const where = ["status = 'published'"];
  const params = {};
  const add = (cond, key, val) => {
    if (val !== undefined && val !== "" && val !== null) {
      where.push(cond);
      params[key] = val;
    }
  };
  add("city = :city", "city", city);
  add("category = :category", "category", category);
  add("kind = :kind", "kind", kind);
  add("listing_type = :listing_type", "listing_type", listing_type);
  add("price_tier = :price_tier", "price_tier", price_tier);
  add("condition_grade = :condition_grade", "condition_grade", condition_grade);
  add("price >= :min", "min", min ? Number(min) : undefined);
  add("price <= :max", "max", max ? Number(max) : undefined);
  if (q) {
    where.push("(title LIKE :q OR description LIKE :q OR brand LIKE :q OR model LIKE :q)");
    params.q = `%${q}%`;
  }
  const page = Math.max(1, Number(req.query.page || 1));
  const limit = Math.min(60, Number(req.query.limit || 24));
  const rows = db
    .prepare(
      `SELECT * FROM listings WHERE ${where.join(" AND ")} ORDER BY created_at DESC LIMIT ${limit} OFFSET ${(page - 1) * limit}`,
    )
    .all(params);
  const total = db.prepare(`SELECT COUNT(*) c FROM listings WHERE ${where.join(" AND ")}`).get(params).c;
  res.json({ items: rows.map(hydrate), total, page, limit });
});

/** GET /listings/mine */
router.get("/mine", requireAuth, (req, res) => {
  const rows = db.prepare("SELECT * FROM listings WHERE owner_id = ? ORDER BY created_at DESC").all(req.user.id);
  res.json({ items: rows.map(hydrate), total: rows.length });
});

/** POST /listings/ — إنشاء مسودة */
router.post("/", requireAuth, (req, res) => {
  const { kind, category, listing_type } = req.body || {};
  const id = uid();
  db.prepare(
    "INSERT INTO listings (id, owner_id, kind, category, listing_type, status) VALUES (?,?,?,?,?,'draft')",
  ).run(id, req.user.id, kind || category || null, category || kind || null, listing_type || null);
  res.status(201).json(hydrate(db.prepare("SELECT * FROM listings WHERE id = ?").get(id)));
});

/** GET /listings/:id */
router.get("/:id", (req, res) => {
  const row = db.prepare("SELECT * FROM listings WHERE id = ?").get(req.params.id);
  if (!row) return res.status(404).json({ detail: "الإعلان غير موجود" });
  db.prepare("UPDATE listings SET views = views + 1 WHERE id = ?").run(row.id);
  res.json(hydrate({ ...row, views: row.views + 1 }));
});

/** POST /listings/:id/update-details */
router.post("/:id/update-details", requireAuth, async (req, res) => {
  const { row, error } = getOwned(req.params.id, req.user);
  if (error) return res.status(error[0]).json({ detail: error[1] });
  const d = req.body?.details ?? req.body ?? {};
  db.prepare(
    `UPDATE listings SET title = COALESCE(?, title), description = COALESCE(?, description),
       city = COALESCE(?, city), district = COALESCE(?, district), area = COALESCE(?, area),
       rooms = COALESCE(?, rooms), bathrooms = COALESCE(?, bathrooms), year = COALESCE(?, year),
       mileage = COALESCE(?, mileage), brand = COALESCE(?, brand), model = COALESCE(?, model),
       details = ?, updated_at = datetime('now') WHERE id = ?`,
  ).run(
    d.title ?? null, d.description ?? null, d.city ?? null, d.district ?? null,
    d.area ?? null, d.rooms ?? null, d.bathrooms ?? null, d.year ?? null,
    d.mileage ?? null, d.brand ?? null, d.model ?? null,
    JSON.stringify(d), row.id,
  );

  // تقدير سعر إرشادي (غير إلزامي)
  const est = await ai.estimatePrice({
    category: row.kind || row.category || "house",
    listing_type: row.listing_type || "sale",
    condition_grade: row.condition_grade || "good",
    features: { ...d, city: d.city ?? row.city },
  });
  if (est?.ok && est.suggested_price) {
    db.prepare("UPDATE listings SET suggested_price = ? WHERE id = ?").run(est.suggested_price, row.id);
  }
  res.json({ ...hydrate(db.prepare("SELECT * FROM listings WHERE id = ?").get(row.id)), price_estimate: est });
});

/** GET /listings/:id/price-suggestion */
router.get("/:id/price-suggestion", requireAuth, async (req, res) => {
  const { row, error } = getOwned(req.params.id, req.user);
  if (error) return res.status(error[0]).json({ detail: error[1] });
  const est = await ai.estimatePrice({
    category: row.kind || row.category || "house",
    listing_type: row.listing_type || "sale",
    condition_grade: row.condition_grade || "good",
    features: json(row.details, {}),
  });
  res.json({ suggested_price: est?.suggested_price ?? row.suggested_price, ...est });
});

/** POST /listings/:id/set-price — بدون سقف إلزامي، فقط تصنيف */
router.post("/:id/set-price", requireAuth, async (req, res) => {
  const { row, error } = getOwned(req.params.id, req.user);
  if (error) return res.status(error[0]).json({ detail: error[1] });
  const price = Number(req.body?.price);
  if (!price || price <= 0) return res.status(422).json({ detail: "أدخل سعراً صحيحاً" });

  let suggested = row.suggested_price;
  if (!suggested) {
    const est = await ai.estimatePrice({
      category: row.kind || row.category || "house",
      listing_type: row.listing_type || "sale",
      condition_grade: row.condition_grade || "good",
      features: json(row.details, {}),
    });
    suggested = est?.suggested_price ?? price;
  }
  const tier = computeTier(price, suggested);
  db.prepare(
    "UPDATE listings SET price = ?, suggested_price = ?, price_tier = ?, updated_at = datetime('now') WHERE id = ?",
  ).run(price, suggested, tier, row.id);
  res.json({ price, suggested_price: suggested, price_tier: tier });
});

/** POST /listings/:id/submit-for-review */
router.post("/:id/submit-for-review", requireAuth, (req, res) => {
  const { row, error } = getOwned(req.params.id, req.user);
  if (error) return res.status(error[0]).json({ detail: error[1] });
  db.prepare("UPDATE listings SET status = 'pending_review', updated_at = datetime('now') WHERE id = ?").run(row.id);
  db.prepare("INSERT INTO notifications (id, user_id, title, body, type, link) VALUES (?,?,?,?,?,?)").run(
    uid(), row.owner_id, "تم إرسال إعلانك للمراجعة", "سيتم إشعارك فور اعتماده.", "listing", `/listing/${row.id}`,
  );
  res.json({ id: row.id, status: "pending_review" });
});

router.delete("/:id", requireAuth, (req, res) => {
  const { row, error } = getOwned(req.params.id, req.user);
  if (error) return res.status(error[0]).json({ detail: error[1] });
  db.prepare("DELETE FROM listings WHERE id = ?").run(row.id);
  res.json({ deleted: row.id });
});
